### PR TITLE
ratio: ratio policy implementation and e2e tests

### PR DESCRIPTION
**Pull request by yanjing1104**

### CODE REVIEW

**Comment by yanjing1104** [comment]
test00-swap and test01-move get passed on both ubuntu-22.04 and debian-sid.

**Comment by yanjing1104** [comment]
@askervin Hi Antti, could you pls help to review it when available? Thank you~~

File: pkg/memtier/policy_ratio.go [hunk 1]
```diff
@@ -0,0 +1,525 @@
...
+					moveSize += addr.length
+				}
+				*idleTcs = append(*idleTcs, (*tcs)[i])
+			}
+		}
+		*tcsList = append(*tcsList, *idleTcs...)
+		log.Debugf("counter b %v moved length: %v, total %v, ratio:%v\n", p.counter, moveSize, memSize, float32(moveSize)/float32(memSize))
+	}
+
+	return tcsList
```
[the remark this answers is not in the document]
Comment (yanjing1104) [reply pkg/memtier/policy_ratio.go:401]: I used "tcs.SortByAccesses()" to sort all the pages first before getting the "ratio" least inactive parts.

File: pkg/memtier/policy_ratio.go [hunk 2]
```diff
@@ -0,0 +1,525 @@
...
+
+// SetConfig() checks the validity of the configuration and sets the policy configuration
+func (p *PolicyRatio) SetConfig(config *PolicyRatioConfig) error {
+	// check "IntervalMs" configuration for ratio policy
+	// optional field, defaults to 20 seconds if missing
+	if config.IntervalMs < 0 {
+		return fmt.Errorf("invalid age policy IntervalMs: %d, > 0 expected", config.IntervalMs)
+	}
+	if config.IntervalMs == 0 {
+		log.Debugf("IntervalMs missing from the ratio policy configuration, defaults to 20 seconds\n")
```
[the remark this answers is not in the document]
Comment (yanjing1104) [reply pkg/memtier/policy_ratio.go]: Done

File: pkg/memtier/policy_ratio.go [hunk 3]
```diff
@@ -0,0 +1,525 @@
...
+
+	// Match "Present" and "Exclusive" memory pages for the given pid
+	pageAttributes, _ := parseOptPages("")
+	pp, err := ar.PagesMatching(pageAttributes)
+	if err != nil {
+		return nil, fmt.Errorf("no address ranges from which to find present/exclusive pages for process %d", pid)
+	}
+
+	// Todo: Obtain the current node based on policy_heat
+	toNode := p.config.RatioTargets[0]
```
[the remark this answers is not in the document]
Comment (yanjing1104) [reply pkg/memtier/policy_ratio.go:272]: Done

File: pkg/memtier/policy_ratio.go [hunk 4]
```diff
@@ -0,0 +1,525 @@
...
+			ppages = ppages.NotOnNode(destNode)
+			if len(ppages.Pages()) > 100 {
+				task := NewMoverTask(ppages, destNode)
+				p.mover.AddTask(task)
+			}
+		}
+	}
+}
+
+// getPagesInfo() returns the pages info of the memory which is not located on the original node(s),
```
[the remark this answers is not in the document]
Comment (yanjing1104) [reply pkg/memtier/policy_ratio.go]: Done

File: pkg/memtier/policy_ratio.go [hunk 5]
```diff
@@ -0,0 +1,525 @@
...
+		int64(swapped)*constPagesize/(1024*1024), int64(pages)*constPagesize/(1024*1024),
+		float32(100*swapped)/float32(pages))
+	return &addrRanges, nil
+}
+
+// getRatioCounters() returns the counters recording the info of the corresponding pages
+// which are the portion based on the ratio configuration of the least active
+func (p *PolicyRatio) getRatioCounters(timestamp int64, Ratio float32) *TrackerCounters {
+	tcsMap := map[int]*TrackerCounters{}
+	memSizeMap := map[int]uint64{}
```
[the remark this answers is not in the document]
Comment (yanjing1104) [reply pkg/memtier/policy_ratio.go]: Done

File: pkg/memtier/policy_ratio.go [hunk 6]
```diff
@@ -0,0 +1,525 @@
...
+			tcsMap[pid] = tcs
+		} else {
+			*tcsMap[pid] = append(*list, *tcs...)
+		}
+	}
+
+	tcsList := &TrackerCounters{}
+	for pid, tcs := range tcsMap {
+		var moveSize uint64
+		memSize := memSizeMap[pid]
```
[the remark this answers is not in the document]
Comment (yanjing1104) [reply pkg/memtier/policy_ratio.go]: Done

File: pkg/memtier/policy_ratio.go [hunk 7]
```diff
@@ -0,0 +1,525 @@
...
+		tcs.SortByAccesses()
+		idleTcs := &TrackerCounters{}
+
+		if p.config.RatioTargets == nil || len(p.config.RatioTargets) == 0 || p.config.RatioTargets[0] == int(NODE_SWAP) {
+			swapout, err := p.getAddrRangesInSwapSpace(pid)
+			if err != nil {
+				log.Errorf("getAddrRangesInSwapSpace error %v", err)
+				continue
+			}
+			moveSize += uint64(len(swapout.Ranges()))
```
[the remark this answers is not in the document]
Comment (yanjing1104) [reply pkg/memtier/policy_ratio.go]: Done

File: pkg/memtier/policy_ratio.go [hunk 8]
```diff
@@ -0,0 +1,525 @@
...
+		} else {
+			pages, err := p.getPagesInNodeSpace(pid)
+			if err != nil {
+				log.Errorf("getPagesInNodeSpace error %v", err)
+				continue
+			}
+			moveSize += uint64(len(pages.pages))
```
[the remark this answers is not in the document]
Comment (yanjing1104) [reply pkg/memtier/policy_ratio.go]: Done

File: pkg/memtier/policy_ratio.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (askervin) [reply pkg/memtier/policy_ratio.go:401]: Indeed I missed it. Thanks, @yanjing1104!

**Comment by askervin** [comment]
Great! Thank you for new policy, @yanjing1104!

There is one thing that I'd like to get optimized in a follow-up PR. That's how to calculate how many bytes have been swapped out. There we currently build a list of address ranges (start and end addresses) that are actually overlapping and there are as many ranges in the list as there are swapped out pages. But I guess all we need is only the number of swapped out pages that you already have in the `swapped` variable. This will save a lot of memory, time and complexity in the calculation.

**Comment by yanjing1104** [comment]
> Great! Thank you for new policy, @yanjing1104!
> 
> There is one thing that I'd like to get optimized in a follow-up PR. That's how to calculate how many bytes have been swapped out. There we currently build a list of address ranges (start and end addresses) that are actually overlapping and there are as many ranges in the list as there are swapped out pages. But I guess all we need is only the number of swapped out pages that you already have in the `swapped` variable. This will save a lot of memory, time and complexity in the calculation.

Yeah sure, will do and thanks for pointing it out again. (Sry I forgot it in this PR, :D)

**Pull request merged**